pkg/discovery/dns/provider.go

```diff
@@ -120,6 +123,7 @@ func (p *Provider) Resolve(ctx context.Context, addrs []string) {
 		resolved, err := p.resolver.Resolve(ctx, name, QType(qtype))
 		p.resolverLookupsCount.Inc()
 		if err != nil {
+			errorList.Add(err)
 			// The DNS resolution failed. Continue without modifying the old records.
 			p.resolverFailuresCount.Inc()
 			level.Error(p.logger).Log("msg", "dns resolution failed", "addr", addr, "err", err)
```

Comment:
Ok this wrong. 

The problem is that we handle errors TWICE. One time here, then we pass `errs` somewhere. This will mean that we probably will log errors couple of times and just add confusion... Same with metric which is even more risky. Let's handle error just once. Either handle it all (just log), or pass further... not both! (: 

What do you think?

Reply:
I think that we should not log error in the `dns/provider.go`, rather the error should be logged by the function requesting it, so that we can clearly log where the error is coming from.

Reply:
If we log, and we do that only for this error, do we really need to return this error? Does the caller ("requesting function") needs to act in specific way for those errors?

Reply:
Okay, so I forgot to mention that, I am not logging the error _now_ in `provider.go`, rather I am returning it(in my current change that I will push), and the receiver would handle that error(by logging it).

Reply:
> Does the caller ("requesting function") needs to act in specific way for those errors?

If we can log in the scope of the requesting function, that would be helpful in understanding which function request resulted in the error. Other than that, there is no specific behaviour that I am looking forward to.

Reply:
I understand this, but still what would be the solid answer to the question: 

> If we log, and we do that only for this error, do we really need to return this error? Does the caller ("requesting function") needs to act in specific way for those errors?

(: ?

Reply:
> If we log, and we do that only for this error, do we really need to return this error? Does the caller ("requesting function") needs to act in specific way for those errors?

As I have understood this question, I think we should return the error and the requesting function should log the error. If you have found any function which is doing both, then please highlight that, I might have missed that :sweat_smile: